icu4c/source/common/uloc.cpp

```diff
@@ -848,20 +848,59 @@ uloc_setKeywordValue(const char* keywordName,
                      const char* keywordValue,
                      char* buffer, int32_t bufferCapacity,
                      UErrorCode* status)
+{
+    if (U_FAILURE(*status)) {
+        return -1;
+    }
+
+    if (bufferCapacity <= 1) {
+        *status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+
+    int32_t bufLen = (int32_t)uprv_strlen(buffer);
+    if(bufferCapacity<bufLen) {
+        /* The capacity is less than the length?! Is this NUL terminated? */
+        *status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+
+    char* keywords = const_cast<char*>(locale_getKeywordsStart(buffer));
+    int32_t baseLen = keywords == nullptr ? bufLen : keywords - buffer;
+    CheckedArrayByteSink sink(keywords == nullptr ? buffer + bufLen : keywords,
+                              bufferCapacity - baseLen - 1);
+    int32_t reslen = ulocimp_setKeywordValue(
+            keywords, keywordName, keywordValue, sink, status);
+
+    if (U_FAILURE(*status)) {
+        return reslen > 0 ? reslen + baseLen : reslen;
+    }
+
+    // See the documentation for this function, it's guaranteed to never
+    // overflow the buffer but instead abort with BUFFER_OVERFLOW_ERROR.
+    U_ASSERT(!sink.Overflowed());
```

Comment:
Why this assertion?
Yes, we don't want to write beyond the capacity, but rather than asserting that the result cannot be too long, we should check for that and return with an overflow error and a length of `baseLen + sink.NumberOfBytesAppended()` right?

Reply:
We know that `ulocimp_setKeywordValue()` must promise to never cause an overflow so this assertion is there to verify that this promise is being kept (and give a nice clear error message in case there's a bug that causes an overflow that should never happen).

Reply:
Sorry, I don't follow. If the new function writes too much to the sink, then the CheckedArrayByteSink here will report Overflowed() but will not have written past the end of the array. When this happens, we need to set a buffer-overflow-error (if there isn't a failure UErrorCode already) and return with the length of the full output. What am I missing?

Reply:
This, on the other hand, is documented:

> NOTE: Unlike almost every other ICU function which takes a buffer, this function will NOT truncate the output text, and will not update the buffer with unterminated text setting a status of U_STRING_NOT_TERMINATED_WARNING. If a BUFFER_OVERFLOW_ERROR is received, it means a terminated version of the updated locale ID would not fit in the buffer, and the original buffer is untouched. This is done to prevent incorrect or possibly even malformed locales from being generated and used.

In order to preserve that behaviour, nothing at all gets written to the sink if there isn't room to write everything that needs to be written (and therefore no overflow can happen).

Reply:
I see, so the sink does not see anything which means it can't have Overflowed().